### ISSUE TITLE
How to weight loss function

### ISSUE THREAD

**Issue by freecui** [huggingface/setfit#375]
I have  an extreme imbalanced dataset for muticlass task . To handle the imbalance, I want to weight loss function

**Comment by kgourgou** [comment]
How extreme? Does it help to pass an option to the head model, i.e., LogisticRegression? For example, you can pass `class_weight='balanced'`, I have found this to work well.

**Comment by freecui** [comment]
from sentence_transformers.losses import CosineSimilarityLoss

from setfit import SetFitModel, SetFitTrainer
from sklearn.linear_model import LinearRegression
from sentence_transformers import SentenceTransformer
import torch
# Choose a sentence transformer for your embedding
model_body = SentenceTransformer("sentence-transformers/paraphrase-mpnet-base-v2")
cls_weights=(1-(train_dataset["label"].value_counts().sort_index()/len(train_dataset))).values
# Choose whatever head you want for regression
model_head = LogisticRegression(class_weight=torch.from_numpy(cls_weights).float())

# Load a SetFit model from Hub
model = SetFitModel(model_body=model_body, model_head=model_head,)

# Create trainer
trainer = SetFitTrainer(
    model=model,
    train_dataset=train_dataset,
    eval_dataset=eval_dataset,
    loss_class=CosineSimilarityLoss,
    metric="accuracy",
    batch_size=16,
    num_iterations=20, # The number of text pairs to generate for contrastive learning
    num_epochs=1, # The number of epochs to use for constrastive learning
    column_mapping={"sentence": "text", "label": "label"} # Map dataset columns to text/label expected by trainer
)

# Train and evaluate
trainer.train()



Is that right?

**Comment by freecui** [comment]
an extreme imbalanced dataset： for example  total data is 30000. class A has 12000, class B has 1, class C has 2000, class D has 13 
 ...

**Comment by freecui** [comment]
I have tried class_weight='balanced', but the result is just 74.57%. If I don't use class_weight='balanced, the result is 80%
the code:
model_body=SentenceTransformer("uer/sbert-base-chinese-nli")
model_head=LogisticRegression(class_weight="balanced")
model = SetFitModel(model_body=model_body,model_head=model_head)
# Create trainer
trainer = SetFitTrainer(
    model=model,
    train_dataset=train_ds,
    eval_dataset=test_ds,
    loss_class=CosineSimilarityLoss,
    batch_size=8,
    num_iterations=10, # Number of text pairs to generate for contrastive learni                                                                                                                                               ng
    num_epochs=1# Number of epochs to use for contrastive learning,
)

trainer.train()

**Comment by Ulipenitz** [comment]
Maybe usefull for people working on a **multi-label** problem:
You can wrap _**LogisticRegression**_ inside _**OneVsRestClassifier**_ like this:

```
model = SetFitModel(
                    model_body=SentenceTransformer('all-MiniLM-L6-v2'), 
                    model_head=OneVsRestClassifier(LogisticRegression(class_weight="balanced")),
                    multi_target_strategy="one-vs-rest"
                    )
```

**Comment by Matthieu-Tinycoaching** [comment]
@freecui @kgourgou where to specify `LogisticRegression(class_weight="balanced")` during the hyperparameters optimization phase :

**Comment by kgourgou** [comment]
@Matthieu-Tinycoaching I'm only guessing, but "head_params" looks like the place to try.

**Comment by Matthieu-Tinycoaching** [comment]
@kgourgou that's what I've done experimentally also. Didn't generate any error when launching, but don't know if this works...
